### ISSUE TITLE
Updating cluster sources broken

### ISSUE THREAD

**Issue by daumann** [uber/react-map-gl#482]
You can reproduce this error by updating a geojson **source** which has the 
```
"markers": {
      "cluster": true,
      "clusterMaxZoom": 14,
      "clusterRadius": 50,
      ...
```
property with for example:

```
     mapStyle.setIn(['sources', sourceId, 'data', 'features'], [{
        "geometry": {
          "coordinates": [
            2.15888889,
            41.36722222
          ],
          "type": "Point"
        },
        "type": "Feature"
      }])
```

Here is the stack trace:

> Uncaught (in promise) Error: There is already a source with this ID
>     at r.addSource (mapbox-gl.js:381)
>     at e.addSource (mapbox-gl.js:495)
>     at updateStyleSource (style-utils.js:224)
>     at setDiffStyle (style-utils.js:107)
>     at StaticMap._updateMapStyle (static-map.js:212)
>     at StaticMap.componentWillReceiveProps (static-map.js:128)
>     at ReactCompositeComponent.js:608
>     at measureLifeCyclePerf (ReactCompositeComponent.js:73)
>     at ReactCompositeComponentWrapper.updateComponent (ReactCompositeComponent.js:607)
>     at ReactCompositeComponentWrapper.receiveComponent (ReactCompositeComponent.js:544)

When testing an earlier version ( 3.0.4) it worked so this problem seems to be recently introduced.

_Somewhat related question: I noticed the [supercluster ](https://www.npmjs.com/package/supercluster)dependency in some of the examples but cannot find any code that is using it?_

**Comment by Pessimistress** [comment]
Is it only a problem if you use cluster?

**Comment by daumann** [comment]
Yes

**Comment by cl0ckwork** [comment]
I ran into this issue as well, her's a snippet of a comment I wrote about it:

if you change some of the extra props (https://www.mapbox.com/mapbox-gl-js/style-spec/#sources-geojson) AND a layer referencing the source is loaded, there will be an error:
this is the culprit 
`/react-nap-gl/style-utils.js line 104: (newSource.clusterRadius === undefined ||newSource.clusterRadius === oldOpts.superclusterOptions.radius)`

`oldOpts.superclusterOptions.radius` is always 800, regardless of the prop passed.

so changing it in your source at the component / reducer level wont trickle down as a change. sorry this was a quick post. hope it helps!

**Comment by Pessimistress** [comment]
@daumann I cannot reproduce your error. It seems that your mapbox does not like this:
```js
map.removeSource(update.id);
map.addSource(update.id, newSource);
```
At https://github.com/uber/react-map-gl/blob/master/src/utils/style-utils.js#L115
Can you double check your react-map-gl and mapbox-gl versions?

There is indeed an issue with what @cl0ckwork reports, which results in removing/adding a source unnecessarily (flicker).

**Comment by Pessimistress** [comment]
Partial fix published v3.2.6
